homeassistant/components/roborock/number.py

```diff
@@ -107,6 +108,12 @@ def native_value(self) -> float | None:
 
     async def async_set_native_value(self, value: float) -> None:
         """Set number value."""
-        await self.entity_description.update_value(
-            self.get_cache(self.entity_description.cache_key), value
-        )
+        try:
+            await self.entity_description.update_value(
+                self.get_cache(self.entity_description.cache_key), value
+            )
+        except RoborockException as err:
+            raise HomeAssistantError(
+                translation_domain=DOMAIN,
+                translation_key="update_options_failed",
+            ) from err
```

Comment:
If this is a common pattern maybe we could make a decorator?

Reply:
I can see where this makes less code in the caller.

I believe this would need to have 3 parameters: The original exception, the domain, and the translation_key (at a minimum, tough there are also additional fields.

In the case in this PR, I used the same translation key. Each entity description has a lambda and to use a decorator that lambda would be decorated:
```
    RoborockNumberDescription(
        key="volume",
        translation_key="volume",
        ...
        @home_assistant_exception_wrapper(RoborockException, DOMAIN, "update_options_failed")
        update_value=lambda cache, value: cache.update_value([int(value)]),
    )
```
and it would be repeated for each description. 

They could also share a common update function `cache_update_value` to avoid copying the decorator to each lambda, however, then they share the same code so a decorator might not be as beneficial within this integration, but could still be useful across integrations.

The other case used in roborock is where the exception is parameterized based on the type of request sent:
```
            raise HomeAssistantError(
                translation_domain=DOMAIN,
                translation_key="command_failed",
                translation_placeholders={
                    "command": command_name,
                },
            ) from err
```
So it probably would not use a decorator since the arguments to the decorator are dependent on the calling code. 

Once you start having multiple exception types caught at once and converted to different types of `HomeAssistantError` then i think the decorator might start to attempt to re-create try/except/except/except/finally semantics.

I'm not a decorator expert so maybe there are better ways to do this, but that was as far as I could get thinking about it. I lean against it. If we'd like to make this PR use less code, however, then an additional method shared by all the entities/entity descriptions would be the way I'd go. I'm OK with the code duplication in this case given it's simple to read is tested.

Reply:
Honestly, I think this is fine 👍